### PR TITLE
[DNM] Peer-to-peer shuffle design

### PR DESCRIPTION
**Pull request by gjoseph92**
Here's a work-in-progress proposal for how we can implement the scalable peer-to-peer shuffling prototyped in https://github.com/dask/dask/pull/8223 in a maintainable and stable way.

The primary problem with the code in https://github.com/dask/dask/pull/8223 is that it's brittle and non-maintainable in its current state. (And also not reliable.) So the primary purpose of this PR is to discuss how we can write something similar, implementation-wise, but in a way that integrates better and we're happy to maintain long-term.

This is meant to be relatively high-level and architectural, and we'd like to keep the conversation away from implementation details except where necessary.

You can view the rendered markdown at https://github.com/gjoseph92/distributed/blob/p2p-shuffle/proposal/distributed/shuffle/shuffle-design.md.

cc @fjetter @jcrist @crusaderky @jrbourbeau

### CODE REVIEW

**Comment by mrocklin** [comment]
In general the objectives and constraints laid out here seems sensible to me.  Thank you for writing this up @gjoseph92 .  I would like to suggest that we think about a plan for resilience, but implement that last.  There is value in getting this out quickly so that people can experiment and give us feedback.  I expect resilience to be an interesting but time consuming and ultimately separable problem.

File: distributed/shuffle/shuffle-design.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+1. Improve performance with concurrency if possible
+1. Spill-to-disk
+1. Backpressure
+1. Performance
+
+## Design
+
+Here we'll discuss the highest-level architectural components of the shuffling system.
+
+_Note: unlike the POC PR, we propose keeping this code in `dask/distributed`, not `dask/dask`. The implementation will be somewhat dependent on worker internals, so internal changes in `distributed` are far more likely to break things than changes in `dask`. We'd like tests to fail (and fixes to happen) in the same repo as the changes that break things. Plus, the code is very `distributed`-specific anyway._
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:45]: > Put the shuffling related tests in importable test classes and import that tests uite in dask/distributed

We're planning on doing this option

File: distributed/shuffle/shuffle-design.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+* Holding per-worker out-of-band state for an in-progress shuffle
+* Adding new handlers in an organized way for workers to transfer shuffle data
+* Doing the above cleanly with multiple concurrent shuffles (`merge`)
+* Coordinating multiple concurrent shuffles which may need to share limited resources (memory, threads, etc.)
+* Getting metrics back to the scheduler/dashboard, like managed memory & bytes spilled to disk (eventually)
+
+The `ShuffleExtension` will be built into distributed and added to workers automatically (like the `PubSubWorkerExtension`). It'll add a route to the worker; something like:
+
+```python
+def shuffle_receive(comm, shuffle_id: str, data: DataFrame) -> None:
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:71]: We may actually want a static ID when we have the `RerunGroup` system. A `RerunGroup` will have a name; when it needs to be rerun, a signal will be sent to all workers keyed by that group name. The extension will listen to that signal and clean up the state when it happens, so reusing the same name should work fine.

File: distributed/shuffle/shuffle-design.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+The `transfer` tasks will pass their input partitions into the `ShuffleExtension`, blocking the task until the extension is ready to receive another input. Internally, the extension will do whatever it needs to do to transfer the data, using worker comms to call the `shuffle_receive` RPC on peer workers. Simultaneously, it'll handle any incoming data from other workers.
+
+### Retries and cancellation
+
+Problems to solve:
+* Causing all tasks in the shuffle to rerun when a worker leaves
+* Cleaning up out-of-band state when a user cancels a shuffle, or it errors
+
+Because most the tasks in the shuffle graph are impure and run for their side effects, restarting an in-progress shuffle requires rerunning _every_ task involved, even ones that appear to have successfully transitioned to `memory` and whose "results" are stored on non-yet-dead workers.
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:97]: > The reason why we'd need to rerun all tasks is because of the hashing, isn't it?

Yes. If the number or order of workers has changed, then the hashing will make the same data go to a different worker the second time around. Your hypothetical is correct.

File: distributed/shuffle/shuffle-design.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+Additionally, cleanly stopping a running shuffle takes more than just releasing the shuffle tasks from memory: since there's out-of-band processing going on, the `ShuffleExtension` has to be informed in some way that it needs to stop doing whatever it's doing in the background, and clear out its buffers. Also, executing tasks may be blocking on the `ShuffleExtension` doing something; without a way to tell the extension to shut down, those tasks might block forever, deadlocking the cluster.
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:99]: > Remove the TaskState.key from Worker.data if it is in there.

In the POC we used this as the signal to clean up state (the `ShuffleService` object was a dependency of every task, got copied into every `Worker.data`, and had a `__del__` method to clean things up when it was removed). 

But this is unreliable in pathological shuffle cases; for example when some workers either don't send any data or don't receive any data in the shuffle. Also, there are weird questions around what happens when that object itself is spilled to disk.

> Whatever the extension does, it might be coupled to any of these mechanisms and a "release task" action would be sufficient for cleanup

I agree that this would be nice, but I don't really like the idea of hooking into `gather_dep` / TaskState attribute changes to detect when this has happened. It just feels brittle to me and like it requires really really careful reasoning about what signals the scheduler is going to send in which cases. Having an explicit mechanism built into `RerunGroup` seems both easy and reliable.

In the spirit of what I said in the blog post:

> It may be worth it to adjust the system so that bypassing task graphs isn’t a hack, but a supported (and tested and maintained) feature.
https://coiled.io/blog/better-shuffling-in-dask-a-proof-of-concept/

I just prefer the idea of making a clear path for how to this in the API that doesn't require deep understanding of the current implementation of the worker and scheduler.

File: distributed/shuffle/shuffle-design.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+Therefore, we propose adding a `RerunGroup` (`ImpureGroup`? `CoExecutionGroup`? `RestartGroup`? `OutOfBandGroup`? name TBD) structure to the scheduler which intertwines the fates of all tasks within it: if any one task is to be rescheduled (due to its worker leaving), all tasks are restarted; if any one is to be prematurely released (due to cancellation), all are released.
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md]: added xref to https://github.com/dask/distributed/issues/5403

File: distributed/shuffle/shuffle-design.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+Membership in a `RerunGroup` is implemented via task annotations, where each task gives the name of the `RerunGroup` it belongs to. A task can belong to at most one `RerunGroup`. TBD if we will enforce any structural restrictions on `RerunGroup`s to prevent odd/invalid states from emerging—we probably should, such as not allowing disjoint parts of the graph to be in the same `RerunGroup`, etc.
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md]: > I'm inclined to say this is out of scope and the responsibility of the user.

I think this could create all sorts of headaches for us down the line in the scheduler, where we have to bend ourselves into knots to handle weird cases that are technically possible but nobody should actually do. A more restrictive API up front means less work for us!

[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md]: > Question: What happens if you fuse an annotated blockwise with an unannotated blockwise. The intended behaviour is currently unclear.

There's an open issue around annotations getting lost during optimization https://github.com/dask/dask/issues/7036 but I don't think it fully recognizes the problem (we don't have a clear policy on annotation propagation during optimization). Fixing this would also be necessary work here; I've mentioned it in the doc.

File: distributed/shuffle/shuffle-design.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+Additionally, the scheduler informs workers whenever a `RerunGroup` is restarted or cancelled. Workers will have a way to pass this information on to any interested out-of-band operations. This could be something like:
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:105]: See my response to https://github.com/dask/distributed/pull/5435/files#r738438974. I think trying to infer that this has happened without an explicit API is even more complicated than adding one.

File: distributed/shuffle/shuffle-design.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+- A `register_cancellation_handler(rerun_group: str, async_callback: Callable)` method on workers that registers an async function to be run when that group is cancelled/restarted. A potential upside (and potential deadlock) is that the scheduler's `cancel_rerun_group` RPC to workers could block on this callback completing, meaning the scheduler wouldn't treat the `RerunGroup` as successfully cancelled until every callback on every worker succeeded. That could give us some nice synchronization guarantees (which we may or many not actually need) ensuring a shuffle doesn't try to restart while it's also trying to shut down.
+
+### Peer discovery and initialization
+
+Problems to solve:
+* Workers need to all have the same list of peers participating in the shuffle (otherwise data could end up in two different places!)
+* Scheduler needs to be told where to run the `unpack` tasks which bring data back in-band
+
+We'll run a single `shuffle-setup` task before all the transfers to do some initialization.
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:115]: > Why does this need to be a single task?

Because it needs to run exactly once. The main thing this task is doing is getting the list of current workers from the scheduler. If we did this in every task, and workers were joining the cluster at the same time, we'd have different lists on different workers and be sending the data to inconsistent places.

There's a way we could do this instead with a scheduler plugin (register the shuffle ID on the scheduler; the first call stores the list of workers and subsequent calls always return the same one) but adding another component to me seems more complicated, since we then need to manage the lifecycle on that state too.

File: distributed/shuffle/shuffle-design.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+First, it will ask the scheduler for the addresses of all workers that should participate in the shuffle (taking into account worker or resource restrictions). How this will be implemented is TBD.
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:117]: But what would trigger that broadcast happening? As I said above, it just seems simpler to coordinate this from a task.

File: distributed/shuffle/shuffle-design.md [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+1. Backpressure
+1. Performance
+
+## Design
+
+Here we'll discuss the highest-level architectural components of the shuffling system.
+
+_Note: unlike the POC PR, we propose keeping this code in `dask/distributed`, not `dask/dask`. The implementation will be somewhat dependent on worker internals, so internal changes in `distributed` are far more likely to break things than changes in `dask`. We'd like tests to fail (and fixes to happen) in the same repo as the changes that break things. Plus, the code is very `distributed`-specific anyway._
+
+### Task graph
```
Comment (gjoseph92) [inline distributed/shuffle/shuffle-design.md:47]: This design breaks with code like:
```python
df2 = df.set_index("name").loc["Alice":"Bob"]
```
We sub-select just the partitions containing rows Alice through Bob. Graph culling removes the `shuffle_unpack` tasks producing the other output partitions:
cull
 
The problem is, our design is highly dependent on knowing how many output partitions will actually be retrieved from the service. The number of output partitions is hardcoded into the graph when we construct it. Each ShuffleExtension tracks how many times `get_output_partition` has been called. When that's happened the right number of times, the extension knows the shuffle is over, and it can clean up.

If not all of the output partitions are actually retrieved, the extension will never clean up, and all that extra data will just sit on the worker forever.

Additionally, we're missing an optimization opportunity: if we only need partitions 1-3, then the extension can drop all inputs that don't match, vastly reducing the amount of data transferred.

_But_, that opens up another problem. What if you do:
```python
shuffled = df.set_index("name")
shuffled.loc["Alice":"Bob"].persist()
sleep(10)
shuffled.persist()
```
First, you only request a few outputs (`(shuffle_unpack-abcde, 0), (shuffle_unpack-abcde, 1)`). The service starts transferring only data for those outputs. Then later, you request more outputs (`(shuffle_unpack-abcde, 0), (shuffle_unpack-abcde, 1), ... (shuffle_unpack-abcde, 100)`).

Critically, _both shuffle graphs have the same dask keys_. From the scheduler's perspective, the `(shuffle_transfer-abcde, 0)` that it already ran is in `"memory"` and doesn't need to run again. But in reality, that task dropped a lot of input data we just now discover we need. The tasks are impure.

Ideally, if the `shuffle-transfer` tasks are going to produce different data, they should have different keys. The most logical thing to do would be rewrite the graph when the `cull` optimization happens, since that's what produces the issue in the first place. We could have an optimization pass that checks how many output tasks are actually needed from the shuffle, inserts this number into the graph, and re-hashes all the tokens.

And just for comparison, what if we don't do this only-send-needed-data optimization? Well, then we have to figure out that the shuffle is over, even if all the data hasn't been collected. I think this is pretty much impossible for an individual worker to detect without talking to all its peers.

So we'd have to have a scheduler plugin to detect this—which, as I mention in the comment below, we probably have to have anyway. So for now, that might be the route forward. But this optimization seems quite important and useful to users, so we should think about the graph-rewriting as well.

File: distributed/shuffle/shuffle-design.md [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+Additionally, the scheduler informs workers whenever a `RerunGroup` is restarted or cancelled. Workers will have a way to pass this information on to any interested out-of-band operations. This could be something like:
+- Workers have a named `threading.Event` for each `RerunGroup` that any of their current tasks belong to. When the scheduler tells workers about a restart/cancellation, they `set()` the corresponding event so that some background thread can respond accordingly.
+- A `register_cancellation_handler(rerun_group: str, async_callback: Callable)` method on workers that registers an async function to be run when that group is cancelled/restarted. A potential upside (and potential deadlock) is that the scheduler's `cancel_rerun_group` RPC to workers could block on this callback completing, meaning the scheduler wouldn't treat the `RerunGroup` as successfully cancelled until every callback on every worker succeeded. That could give us some nice synchronization guarantees (which we may or many not actually need) ensuring a shuffle doesn't try to restart while it's also trying to shut down.
+
+### Peer discovery and initialization
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/shuffle-design.md:109]: > Say you have 100 workers and 2 shuffles. The two shuffle_inits run on workers 0 and 1. Then worker 99 leaves. shuffle_init would not be rescheduled, because no keys in the RerunGroup were lost.

That assumes that the `RerunGroup` triggers on released key. We could just as well trigger it on `remove_worker` and there is no problem.

> We should probably suck it up and create a scheduler plugin for shuffling.

Indeed this is something I would've started with. I don't consider it an additional layer of complexity since it simplifies the graph. However, we still need to do the same thing. Whether we put this into the shuffle-setup, the scheudler plugin, the barrier, etc. we still need to 
* generate a unique shuffle id
* release all keys if something goes wrong (e.g. error, worker lost)
* set output restrictions
* ???
In my mind the only new (is it actually new??) requirement is to (trigger a `RerunGroup` | restart a shuffle) if a worker is lost. The plugin is a different way to do it and it might have synergies with other requirements (e.g. if we can avoid another complex optimization step and deal with culling more sensibly, I'm all in favour of the plugin)

I wouldn't reject the idea of a `RerunGroup` entirely based on this. However, I see the appeal of a SchedulerPlugin

Re: barrier task

> [Excalidraw] no-op dummy task to create a dependency between `unpack` and `transfer` and enforce task ordering. In theory, we could just have an all-to-all graph like this [...] but that would be much more burden on the scheudler to handle.

Just a technical comment since I was thrown off by this as well for a second. This is not a huge burden for the scheduler. The scheduler doesn't care (a lot) about the edges we have in our graph.
Edges for the scheduler merely mean
* dask.order is computationally more expensive (that might be sizeable)
* A task transition causes some set.remove/discard but I consider this cheap even for large sets
Otherwise edges mean that the `compute-task` signal for `unpack` includes a list of all workers in the `who_has` key but that's also relatively cheap.

I assume the biggest factor will be that every worker will again need to connect to every worker to fetch the result of every `transfer`. While the result is likely None, the scheduler will still open a comm to every worker. upon receiving this key, it will let the scheduler know about it with `add-keys`. We also need to do all of this with the barrier task but only once and not for every worker.
I would assume all of this only matters for very large clusters otherwise I don't think this is a game changer.
Bottom line, I'm fine with the barrier task and think we should keep it since all of the above would still be unnecessary even if it is not sizeable (imo, we don't need to argue how sizeable it actually is). If at some point it causes problems we should still be open to throw it away again.

File: distributed/shuffle/shuffle-design.md (hunk 10 above)
Comment (fjetter) [reply distributed/shuffle/shuffle-design.md:47]: A different approach to graph rewriting is to not hard code anything into the graph but rather submit these paramters out of band. the `transfer` tasks would get the relevant paramters from the WorkerExtensions instead. That way, the graph can be written agnostically, culled, etc. and everything of relevance is determined at runtime.
That, of course, assumes some kind of synchronization event at the beginning of the shuffle either via a shuffle-setup or via a scheduler plugin

I think to break this problem up a bit, it helps to write down all shuffle input parameters and classify them based on whether the information is required at runtime or at graph construct time.

Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:47]: > A different approach to graph rewriting is to not hard code anything into the graph but rather submit these paramters out of band

I like this way of thinking about it. I'll give that some thought.

Fundamentally though, in order to have the only-send-needed-data optimization, I think we have to rewrite the keys in the graph. Otherwise, you could always have a situation where when you do the full `shuffled.persist()`, the scheduler only decides to schedule some/none of the transfer tasks, because from its perspective those keys are already in memory from `shuffled.loc["Alice":"Bob"].persist()`. Even if all the parameters are fetched at runtime from a scheduler extension, you still need the tasks to get scheduled to actually fetch them.

File: distributed/shuffle/shuffle-design.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle-design.md:109]: > We could just as well trigger it on `remove_worker` and there is no problem.

We could. And I think for shuffling, we actually need to (even without the setup task). As a generic API, it feels a little weird to restart the tasks even when none of the tasks have been affected though. It would be so specially tuned for shuffling at that point, maybe it just shouldn't be a generic API.

> This is not a huge burden for the scheduler.

Good to know. Agreed that it's probably _more_ of a burden than the barrier, but if the difference isn't huge, maybe it doesn't matter. Still, the barrier task doesn't really cause problems currently, so I think I'll probably stick with it for now.
